src/dotnet/RiderPlugin.UnrealLink/PluginInstaller/UnrealProjectsRefresher.cs

```diff
@@ -0,0 +1,255 @@
+                var backgroundTaskHost = solution.GetComponent<RiderBackgroundTaskHost>();
+                backgroundTaskHost.AddNewTask(lifetime, task);
+                var uprojectFile = installDescription?.UprojectPath;
+                if (uprojectFile.IsNullOrEmpty())
+                {
+                    var cppUe4SolutionDetector = solution.GetComponent<CppUE4SolutionDetector>();
+                    using (solution.Locks.UsingReadLock())
+                    {
+                        uprojectFile = cppUe4SolutionDetector.GetUProjectPath();   
+                    }
+                }
+
+                await lifetime.StartBackground(() => RegenerateProjectFiles(lifetime, solution, unrealHost, engineRoot, uprojectFile));
+            });
+        }
+        
+        private static void RegenerateProjectFiles(Lifetime lifetime, ISolution solution, UnrealHost unrealHost,
+            VirtualFileSystemPath engineRoot, VirtualFileSystemPath uprojectFile)
+        {
+            void LogFailedRefreshProjectFiles()
+            {
+                unrealHost.myModel.RiderLinkInstallMessage(new InstallMessage("Failed to refresh project files",
+                    ContentType.Normal));
+                unrealHost.myModel.RiderLinkInstallMessage(
+                    new InstallMessage("RiderLink will not be visible in solution explorer", ContentType.Normal));
+                unrealHost.myModel.RiderLinkInstallMessage(new InstallMessage(
+                    "Need to refresh project files in Unreal Editor or in File Explorer with context action for .uproject file 'Refresh Project files'",
+                    ContentType.Normal));
+            }
+            if (!engineRoot.IsValidAndExistDirectory())
+            {
+                ourLogger.Warn($"[UnrealLink]: Couldn't find Unreal Engine root");
+
+                LogFailedRefreshProjectFiles();
+                return;
+            }
+
+            var pathToUnrealBuildToolBin = CppUE4FolderFinder.GetAbsolutePathToUnrealBuildToolBin(engineRoot);
+
+            // 1. If project is under engine root, use GenerateProjectFiles.{extension} first
+            if (GenerateProjectFilesCmd(lifetime, solution, unrealHost, uprojectFile, engineRoot)) return;
+            // 2. If it's a standalone project, use UnrealVersionSelector
+            //    The same way "Generate project files" from context menu of .uproject works
+            if (RegenerateProjectUsingBundledUVS(lifetime, unrealHost, uprojectFile, engineRoot)) return;
+            if (RegenerateProjectUsingSystemUVS(lifetime, solution, unrealHost, uprojectFile)) return;
+            // 3. If UVS is missing or have failed, fallback to UnrealBuildTool
+            if (RegenerateProjectUsingUBT(lifetime, unrealHost, uprojectFile, pathToUnrealBuildToolBin, engineRoot)) return;
+
+            ourLogger.Warn("[UnrealLink]: Couldn't refresh project files");
+            lifetime.ToCancellationToken().ThrowIfCancellationRequested();
```

Comment:
I don't see a reason to check for cancellation here, the action has tried it's best and failed, the best we can do is show the failed message.